File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.java
class Solution {
    public int[] nextLargerNodes(ListNode head) {

        ListNode curr = head;
        Stack<Integer> st = new Stack<>();

        int size = 0;
        ArrayList<Integer> ans = new ArrayList<>();

        while (curr != null) {
            ans.add(curr.val);
            curr = curr.next;
            size++;
        }

        for (int i = 0; i < size; i++) {

            while (!st.empty() && ans.get(st.peek()) < ans.get(i)) {
                ans.set(st.peek(), ans.get(i));
                st.pop();
            }

            st.push(i);
        }

        while (!st.empty()) {
            ans.set(st.peek(), 0);
            st.pop();
        }

        // Convert ArrayList<Integer> to int[]
        int[] result = new int[size];

        for (int i = 0; i < size; i++) {
            result[i] = ans.get(i);
        }

        return result;
    }
}

File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.cpp
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        ListNode* curr = head;
        stack<int> st;

        int size = 0;
        vector<int> ans;
        while (curr) {
            ans.push_back(curr->val);
            curr = curr->next;
            size++;
        }

        for (int i = 0; i < size; i++) {
            while (!st.empty() && ans[st.top()] < ans[i]) {
                ans[st.top()] = ans[i];
                st.pop();
            }
            st.push(i);
        }

        while (!st.empty()) {
            ans[st.top()] = 0;
            st.pop();
        }

        return ans;
    }
};